File: src/context.ts
import { AsyncLocalStorage } from "node:async_hooks";

type RequestLocalReadResult<T> = { readonly status: "hit"; readonly value: T } | { readonly status: "miss" };

/** @internal */
export class RequestLocalCache {
  readonly inFlight = new Map<string, Promise<unknown>>();
  private readonly values = new Map<string, unknown>();
  private closed = false;

  read<T>(key: string): RequestLocalReadResult<T> {
    if (this.closed || !this.values.has(key)) {
      return { status: "miss" };
    }
    return { status: "hit", value: this.values.get(key) as T };
  }

  set<T>(key: string, value: T): void {
    if (!this.closed) {
      this.values.set(key, value);
    }
  }

  close(): void {
    this.closed = true;
    this.values.clear();
    this.inFlight.clear();
  }
}

interface RequestHolder {
  closed: boolean;
  requestLocalCache?: RequestLocalCache;
}

interface ContextStore {
  readonly enabled: boolean;
  readonly holder: RequestHolder | null;
}

const contextStorage = new WeakMap<DialCacheContext, AsyncLocalStorage<ContextStore>>();

export class DialCacheContext {
  constructor() {
    contextStorage.set(this, new AsyncLocalStorage<ContextStore>());
  }

  isEnabled(): boolean {
    const store = storageFor(this).getStore();
    return store?.enabled === true && store.holder?.closed === false;
  }

  enable<T>(fn: () => T | Promise<T>): Promise<T> {
    const currentHolder = this.liveHolder();
    if (currentHolder !== null) {
      return this.run({ enabled: true, holder: currentHolder }, fn);
    }

    const holder: RequestHolder = { closed: false };
    return this.runOutermost(holder, fn);
  }

  disable<T>(fn: () => T | Promise<T>): Promise<T> {
    return this.run({ enabled: false, holder: this.liveHolder() }, fn);
  }

  private liveHolder(): RequestHolder | null {
    const holder = storageFor(this).getStore()?.holder;
    return holder !== undefined && holder !== null && !holder.closed ? holder : null;
  }

  private async runOutermost<T>(holder: RequestHolder, fn: () => T | Promise<T>): Promise<T> {
    try {
      return await this.run({ enabled: true, holder }, fn);
    } finally {
      holder.closed = true;
      holder.requestLocalCache?.close();
      delete holder.requestLocalCache;
    }
  }

  private async run<T>(store: ContextStore, fn: () => T | Promise<T>): Promise<T> {
    return await storageFor(this).run(store, async () => await fn());
  }
}

/** @internal */
export function getOrCreateRequestLocalCache(context: DialCacheContext): RequestLocalCache | null {
  const store = storageFor(context).getStore();
  const holder = store?.holder;
  if (store?.enabled !== true || holder === null || holder === undefined || holder.closed) {
    return null;
  }

  holder.requestLocalCache ??= new RequestLocalCache();
  return holder.requestLocalCache;
}

function storageFor(context: DialCacheContext): AsyncLocalStorage<ContextStore> {
  return contextStorage.get(context)!;
}

File: python/dialcache/context.py
"""Per-instance request scope with a shared, explicitly closed memo holder."""

from __future__ import annotations

from contextvars import ContextVar, Token
from dataclasses import dataclass, field
from types import TracebackType
from typing import Any


@dataclass
class RequestLocalCache:
    """Request values and flights; closing prevents all late publication."""

    in_flight: dict[str, Any] = field(default_factory=dict)
    _values: dict[str, Any] = field(default_factory=dict)
    closed: bool = False

    def read(self, key: str) -> tuple[bool, Any]:
        if self.closed or key not in self._values:
            return False, None
        return True, self._values[key]

    def set(self, key: str, value: Any) -> None:
        if not self.closed:
            self._values[key] = value

    def close(self) -> None:
        self.closed = True
        self._values.clear()
        self.in_flight.clear()


@dataclass
class _Holder:
    closed: bool = False
    memo: RequestLocalCache | None = None

    def close(self) -> None:
        self.closed = True
        if self.memo is not None:
            self.memo.close()
            self.memo = None


@dataclass(frozen=True)
class _Store:
    enabled: bool
    holder: _Holder | None


class _Scope:
    def __init__(self, context: DialCacheContext, enabled: bool) -> None:
        self._context = context
        self._enabled = enabled
        self._token: Token[_Store | None] | None = None
        self._owned: _Holder | None = None
        self._entered = False

    def __enter__(self) -> _Scope:
        if self._entered:
            raise RuntimeError("A DialCache scope context manager can only be entered once")
        self._entered = True
        holder = self._context._live_holder()
        if self._enabled and holder is None:
            holder = self._owned = _Holder()
        self._token = self._context._storage.set(_Store(self._enabled, holder))
        return self

    def __exit__(
        self,
        exception_type: type[BaseException] | None,
        exception: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        if self._token is None:
            raise RuntimeError("DialCache scope was not entered or was already closed")
        if self._owned is not None:
            self._owned.close()
        self._context._storage.reset(self._token)
        self._token = None

    async def __aenter__(self) -> _Scope:
        return self.__enter__()

    async def __aexit__(
        self,
        exception_type: type[BaseException] | None,
        exception: BaseException | None,
        traceback: TracebackType | None,
    ) -> None:
        self.__exit__(exception_type, exception, traceback)


class DialCacheContext:
    """An independently enabled context for one cache instance.

    Both ``with context.enable():`` and ``async with context.enable():`` are
    supported. Async tasks inherit the live holder, but calls made after its
    outer scope exits are disabled, even from a copied context.
    """

    def __init__(self) -> None:
        self._storage: ContextVar[_Store | None] = ContextVar("dialcache_scope", default=None)

    def _live_holder(self) -> _Holder | None:
        store = self._storage.get()
        if store is None or store.holder is None or store.holder.closed:
            return None
        return store.holder

    def is_enabled(self) -> bool:
        store = self._storage.get()
        return store is not None and store.enabled and self._live_holder() is not None

    def enable(self) -> _Scope:
        return _Scope(self, True)

    def disable(self) -> _Scope:
        return _Scope(self, False)

    def request_cache(self) -> RequestLocalCache | None:
        if not self.is_enabled():
            return None
        holder = self._live_holder()
        assert holder is not None
        if holder.memo is None:
            holder.memo = RequestLocalCache()
        return holder.memo


def get_or_create_request_local_cache(context: DialCacheContext) -> RequestLocalCache | None:
    return context.request_cache()
